main.c

```diff
@@ -265,6 +265,15 @@ int main(void) {
 	app_uartcomm_initialize();
 	app_configuration *appconf = mempools_alloc_appconf();
 	conf_general_read_app_configuration(appconf);
+
+	// check whether we're rebooting from a watchdog reset
+	if (timeout_had_IWDG_reset(false)) {
```

Comment:
I can see this working out poorly if someone expects the system to come back online after a reset, only to find that their app isn't being used. This makes fault tolerance somewhat trickier.

On the flip side, you're dealing with the very real problem of people who cannot access their SWD port and find themselves locked out of their board. Would a happy compromise be to add some more conditions before setting the app to UART? Perhaps checking the number of boot loops vs. how long it was since the last reset (under the assumption that if there is a 60 second window when someone could upload new firmware, that's sufficient to reset the boot loop counter)

Reply:
is there a way to count boot loops and measure time since reset? In my experience the boot loop has a period of just a few seconds, i.e. a subsequent reboot gets triggered right after startup. Are you suggesting to add a WDT reboot counter to persistent memory and then clear it once we've been up for more than x seconds?

In my experience loading firmware requires a lot more time than 60 seconds (bluetooth), but at the same time the reboots I've seen usually happen after less than 10 seconds, so having enough time for a firmware download isn't really important.

Reply:
Is the app thread not terminated when you start a firmware upload? Terminating the app and mcpwm thread on starting a fw upload seems like that may be a good idea. After ensuring the motor is at 0 current.

Reply:
Pretty sure the app isn't terminated. I remember the early version of 5.3 had my motor wiggling (some weird brake current behavior) while firmware was uploading. If I recall correctly I actually submitted a patch to deal with it but it get rejected because Mitch added a different workaround/timeout for the brake behavior